src/unix/uilocale.cpp

```diff
@@ -338,6 +357,14 @@ wxUILocaleImplUnix::Use()
 
         wxSetlocaleTryAll(LC_ALL, locIdAlt);
     }
+
+#ifdef HAVE_LANGINFO_H
+    // In case of the system default locale the codeset can not be
+    // determined properly in the constructor, because the default
+    // locale is actually not set yet.
+    // Therefore we have to do it here (again).
+    m_codeset = GetLangInfo(CODESET);
+#endif
```

Comment:
Sorry, this seems confusing to me: either we don't need it until `Use()` is called and then we shouldn't set it in ctor at all and only set it here or, if we do need it even before `Use()`, we should set it correctly in the ctor for the default locale too, by switching to it first (as we already do now).

But having wrong value of `m_codeset` only for the default locale and only until `Use()` is called just doesn't seem right.

P.S. See also `wxLocale::GetSystemEncodingName()`.

Reply:
> Sorry, this seems confusing to me: either we don't need it until `Use()` is called and then we shouldn't set it in ctor at all and only set it here or, if we do need it even before `Use()`, we should set it correctly in the ctor for the default locale too, by switching to it first (as we already do now).

Unfortunately, the situation is rather complex. Method `GetLangInfo` that is used to determine the codeset depends on `locale_t` support, it uses the current locale if `locale_t` support is not available. And the current locale will not be the right one in the ctor.

> But having wrong value of `m_codeset` only for the default locale and only until `Use()` is called just doesn't seem right.

Hm, it's all about avoiding to tamper with the locale. If we want to determine the correct value in the ctor we need to set the locale temporarily and to reset to the previous locale, although we know that `Use()` will be called for the default locale anyway shortly afterwards. 

Theoretically, the codeset and the name could be determined in method `Use()`. That would avoid most problems. **BUT** we have the special use case of method `GetSystemLocale` where we need to retrieve the locale name **without** calling `Use()`.

> P.S. See also `wxLocale::GetSystemEncodingName()`.

Yes, that code follows variant 1.

I will look into the code again and will try to apply variant 2 - and to fall back to variant 1, if `locale_t `support is missing - for both items, locale name and codeset.

Reply:
Ideal would be to have a (internal) class `TempDefaultLocaleSetter` that could be used to set the given `LC_XXX` to "" in its ctor and back to the original value in its dtor. Then we'd provide 2 implementations of it: one if extended locale support is available, and the other, using `setlocale()`, if it isn't.

Thanks again!

Reply:
> Ideal would be to have a (internal) class `TempDefaultLocaleSetter` that could be used to set the given `LC_XXX` to "" in its ctor and back to the original value in its dtor. Then we'd provide 2 implementations of it: one if extended locale support is available, and the other, using `setlocale()`, if it isn't.

I saw your message just now, after already having modified my local code without using such an internal class. Before I decide to implement such a class (not sure whether it's really worth the effort), we have to further discuss the whole task,

We have 4 use cases:

1. `wxUILocale::UseDefault()`
    Name and codeset can be determined in the ctor, either via `locale_t` support or via `setlocale`. 
2. `wxUILocale::UseLocaleName`
    For `wxLocale` compatibility. If `locale_t` support is available, a matching locale (with UTF-8 codeset, if possible) will be created and determining name and codeset can be done in the ctor. However, without `locale_t` support no locale is created and we would have to duplicate the code from the `Use()` method, where various charset variants (for UTF-8) are tested. And that's why I would prefer to determine the codeset in the `Use()` method, after the locale has been set.
3. `wxUILocale::GetSystemLocale()`
    Can be handled in the same way as `wxUILocale::UseDefault()`.
4. `wxUILocale::wxUILocale(const wxLocaleIdent& localeId)`
    The current locale will not be changed. If `locale_t` support is available this case is similar as `wxUILocale::UseLocaleName`. However, if `locale_t` support is not available, calls to the `GetInfo` method will use the current locale. And that will result in wrong results, unless the `wxUILocale` instance references the same locale as the current locale.

All will work as long as `locale_t` support is available. But if it's not, then use cases 2 and 4 are problematic.